docs/source/dev_environment_setup.md
# Setting up a development environment

## Prerequisites

### Pre-commit hook (for Python code formatting with black)

FlowKit's Python code is formatted with [black](https://black.readthedocs.io/en/stable/) (which is included in the
FlowKit pipenv environments). There is also a [pre-commit](https://pre-commit.com/) hook which runs black on all Python
files before each commit.

To install the pre-commit hook run:
```bash
$ pre-commit install
```
If you ever want to uninstall the hook you can do this by running `pipenv run pre-commit uninstall`.

The above command requires `pre-commit` to be installed on your system. For convenience, it is also
included in the `flowmachine` pipenv environment, so if you don't want to install it system-wide you
can run it as follows (it will still be installed and available for all FlowKit components, not just
flowmachine).
```bash
$ cd flowmachine/
$ pipenv run pre-commit install
```

Note that if you run `git commit` and any files are modified by the re-formatting, the pre-commit hook will abort
the commit (but leave the files re-formatted). Simply repeat the `git commit` command in order to complete the commit.

Example:

```bash
# 1) First attempt to commit; this is aborted because one of the files
#    is reformatted by the pre-commit hook.

$ git commit -a -m "Some changes"
black...................................................Failed
hookid: black

Files were modified by this hook. Additional output:

reformatted flowmachine/__init__.py
All done! ✨ 🍰 ✨
1 file reformatted.


# 2) Complete the commit by running the same 'git commit' command again.

$ git commit -a -m "Some changes"
black...................................................Passed
[black-pre-commit-hook 8ebaace] Some changes
 1 file changed, 2 insertions(+), 1 deletion(-)
```

### Diff tool (for verifying changes in `ApprovalTests`-based tests)

Some of the tests use [ApprovalTests](https://github.com/approvals/ApprovalTests.Python) to verify large output against
a known "approved" version (stored in files called `*approved.txt`). For example, the API specification is
verified in this way. If you make code changes that alter the results of these tests, the content of the
relevant `*.approved.txt` file needs to be updated. ApprovalTests will do this automatically for you, but you need to have a
diff tool installed. (See here for some recommendations for diff tools on [Mac](`https://www.git-tower.com/blog/diff-tools-mac`)
and [Windows](https://www.git-tower.com/blog/diff-tools-windows).)

There is a toplevel config file called `approvaltests_diff_reporters.json` which lists the diff tools that ApprovalTests
will try to find. Currently only `opendiff` is listed there (but we may add additional ones in the future). If you are using a different tool you should
manually add it to this file. See the ApprovalTests [README](https://github.com/approvals/ApprovalTests.Python) for
examples (illustrative particularly for file paths on Windows).

!!! warning
    The project root contains an env file (`development_environment`), in which are default values for _all_ of the environment variables used to control and configure the FlowDB, FlowAPI, FlowMachine and FlowAuth. You will need to source this file (`set -a && . ./development_environment && set +a`), or otherwise set the environment variables before running _any other command_.

## Option 1: Starting up all FlowKit components inside a dockerised development environment

For convenience, FlowKit comes with a dockerised development environment. You can start up a development version of all
components by running:
```bash
$ make up
```

This will use the specifications in `docker-compose.yml` to spin up development versions of `flowdb_testdata`,
`flowmachine`, `flowapi` and `flowauth` (as well as a `redis` container, which is used internally by `flowmachine`).

The first time you run this command it will build the docker images locally, which will take ~15 minutes. Subsequent
runs will be much faster because it will typically only need to (re-)start the containers, or re-build small parts
of the images.

Here is an example of the running docker services after a successful `make up`:
```
$ docker ps

CONTAINER ID        IMAGE                               COMMAND                  CREATED             STATUS              PORTS                              NAMES
54857c5f5da5        flowminder/flowdb-testdata:latest   "docker-entrypoint.s…"   5 minutes ago       Up 5 minutes        8000/tcp, 0.0.0.0:9000->5432/tcp   flowdb_testdata
e57c8eeb7d38        flowminder/flowmachine:latest       "/bin/sh -c 'pipenv …"   5 minutes ago       Up 5 minutes        0.0.0.0:5555->5555/tcp             flowmachine
3d54ce0c4484        flowminder/flowapi:latest           "/bin/sh -c 'pipenv …"   5 minutes ago       Up 5 minutes        0.0.0.0:9090->9090/tcp             flowapi
89e9575c8d42        flowminder/flowauth:latest          "/entrypoint.sh /sta…"   5 minutes ago       Up 5 minutes        443/tcp, 0.0.0.0:8080->80/tcp      flowauth
384355e94ae6        bitnami/redis                       "/entrypoint.sh /run…"   5 minutes ago       Up 5 minutes        0.0.0.0:6379->6379/tcp             query_locker
```

**Note:** _Currently there is a glitch which means that `flowmachine` may not start up correctly if `flowdb_testdata`
is still in the start-up phase while the `flowmachine` container is already up and running. This will be fixed soon but for the
time being you can check if there was an error by checking the `flowmachine` logs (via `docker logs -f flowmachine`),
and if necessary running `make flowmachine-down` followed by `make flowmachine-up`)._

To make changes to the containers, you should run `make down && make up`.


## Option 2: Starting Flowmachine and FlowAPI manually, outside the dockerised setup

While the fully dockerised setup described above is convenient, it has the disadvantage that interactive debugging is
difficult or impossible if Flowmachine and FlowAPI are running inside a docker container. If you want to set breakpoints
and use for example [pdb](https://docs.python.org/3/library/pdb.html), [ipdb](https://pypi.org/project/ipdb/) or similar
debuggers to step through the code then it is easier to start the relevant components manually.

To do this, first start the components which you want to run inside the dockerised environment, e.g. via:
```bash
$ make flowmachine_testdata-up flowauth-up
```

Next, run one or more of the following commands (in separate terminals) to start up the remaining components.

### Flowmachine
```bash
$ cd flowmachine/
$ pipenv run watchmedo auto-restart --recursive --patterns="*.py" --directory="." pipenv run flowmachine
```

### FlowAPI
```bash
$ cd flowapi/
$ pipenv run hypercorn --debug --reload --bind 0.0.0.0:9090 "app.main:create_app()"
```

### FlowAuth

```bash
$ cd flowauth/
$ pipenv run start-all
```

!!! warning
    If you have started FlowMachine and FlowAPI directly, you will need to set the `FLOWKIT_INTEGRATION_TESTS_DISABLE_AUTOSTART_SERVERS=TRUE` before running the integration tests.
    Without this setting, both will be started automatically to allow test coverage to be collected.

# Verifying the setup

This section provides example commands which you can run to verify that `flowdb`, `flowmachine` and `flowapi` started up
successfully and are wired up correctly so that they can talk to each other.

## Running the integration tests

You can run the integration tests as follows. If these pass, this is a good indication that everything is set up correctly.
```bash
$ cd integretion_tests
$ pipenv install  # only needed once to install dependencies
$ pipenv run pytest
```

## FlowDB

```
$ psql "postgresql://flowdb:flowflow@localhost:9000/flowdb" -c "SELECT flowdb_version()"
```
The output should be similar to this:
```
        flowdb_version
-------------------------------
 (0.2.2-8-g0558488,2019-01-21)
(1 row)
```

## Flowmachine

From within the `flowmachine/` folder, run the following command to send an example message to the Flowmachine server via ZeroMQ.
```
$ pipenv run python -c "from flowmachine.core.server.utils import FM_EXAMPLE_MESSAGE, send_message_and_receive_reply; print(send_message_and_receive_reply(FM_EXAMPLE_MESSAGE))"
```
Expected output:
```
Sending message: {'action': 'run_query', 'query_kind': 'daily_location', 'request_id': 'DUMMY_ID', 'params': {'date': '2016-01-01', 'daily_location_method': 'last', 'aggregation_unit': 'admin3', 'subscriber_subset': 'all'}}
{'status': 'accepted', 'id': 'ddc61a04f608dee16fff0655f91c2057'}
```

## FlowAPI

First navigate to `http://localhost:8080/` (where Flowauth should be running), create a valid token (TODO: add
instructions how to do this or link to the appropriate section in the docs) and store it in the environment
variable `TOKEN`.

Then run the following command to submit a daily location calculation to FlowKit via the API:
```
$ curl -v -X POST http://localhost:9090/api/0/run \
  -H "Authorization: Bearer ${TOKEN}" \
  -H 'Cache-Control: no-cache' \
  -H 'Content-Type: application/json' \
  -d '{"params": {"date":"2016-01-01", "level":"admin3", "daily_location_method":"last", "aggregation_unit":"admin3", "subscriber_subset":"all"}, "query_kind":"daily_location"}'
```

The output should contain a location header that similar to this:
```
[...]
< location: /api/0/poll/ddc61a04f608dee16fff0655f91c2057
[...]
```


docs/source/developer/dev_environment_setup.md
# Setting up a development environment

## Option 1: Starting up all FlowKit components inside a dockerised development environment

For convenience, FlowKit comes with a dockerised development environment. You can start up a development version of all
components by running:
```bash
$ make up
```

This will use the specifications in `docker-compose-dev.yml` to spin up development versions of `flowdb_testdata`,
`flowmachine`, `flowapi` and `flowauth` (as well as a `redis` container, which is used internally by `flowmachine`).

The first time you run this command it will build the docker images locally, which will take ~15 minutes. Subsequent
runs will be much faster because it will typically only need to (re-)start the containers, or re-build small parts
of the images.

Here is an example of the running docker services after a successful `make up`:
```
$ docker ps

CONTAINER ID        IMAGE                               COMMAND                  CREATED             STATUS              PORTS                              NAMES
54857c5f5da5        flowminder/flowdb:testdata-latest   "docker-entrypoint.s…"   5 minutes ago       Up 5 minutes        8000/tcp, 0.0.0.0:9000->5432/tcp   flowdb_testdata
e57c8eeb7d38        flowminder/flowmachine:latest       "/bin/sh -c 'pipenv …"   5 minutes ago       Up 5 minutes        0.0.0.0:5555->5555/tcp             flowmachine
3d54ce0c4484        flowminder/flowapi:latest           "/bin/sh -c 'pipenv …"   5 minutes ago       Up 5 minutes        0.0.0.0:9090->9090/tcp             flowapi
89e9575c8d42        flowminder/flowauth:latest          "/entrypoint.sh /sta…"   5 minutes ago       Up 5 minutes        443/tcp, 0.0.0.0:8080->80/tcp      flowauth
384355e94ae6        bitnami/redis                       "/entrypoint.sh /run…"   5 minutes ago       Up 5 minutes        0.0.0.0:6379->6379/tcp             redis_flowkit
```

**Note:** _Currently there is a glitch which means that `flowmachine` may not start up correctly if `flowdb_testdata`
is still in the start-up phase while the `flowmachine` container is already up and running. This will be fixed soon but for the
time being you can check if there was an error by checking the `flowmachine` logs (via `docker logs -f flowmachine`),
and if necessary running `make flowmachine-down` followed by `make flowmachine-up`)._

In this dev setup, the `flowmachine/` and `flowapi/` source folders are mounted as volumes inside the `flowmachine`
and `flowapi` containers. In addition, the `flowmachine` and `flowapi` services are being run in "live reload" mode
(in `flowmachine`'s case this is achieved via [watchdog](https://pythonhosted.org/watchdog/)).
This ensures that any changes to the source code automatically trigger a restart of the process and there is no need to
re-build the images or re-start the docker containers when changes to the code are made during development. (The
only exception is if new dependencies are added, in which case you can run `make down` followed by `make up` to
re-build the images and re-start the containers.)

_(Note: currently this live reload is not enabled for flowauth, but it is planned and will be available very shortly.)_

Any changes to `flowdb` - for example, tweaks to the database schema - will require a re-build and a re-start of the
`flowdb` and/or `flowdb_testdata` image and a re-start of the container. Simply run `make down` followed by `make up`
to achieve this.


## Option 2: Starting Flowmachine and FlowAPI manually, outside the dockerised setup

While the fully dockerised setup described above is convenient, it has the disadvantage that interactive debugging is
difficult or impossible if Flowmachine and FlowAPI are running inside a docker container. If you want to set breakpoints
and use for example [pdb](https://docs.python.org/3/library/pdb.html), [ipdb](https://pypi.org/project/ipdb/) or similar
debuggers to step through the code then it is easier to start the relevant components manually.

To do this, first start the components which you want to run inside the dockerised environment, e.g. via:
```bash
$ make flowmachine_testdata-up flowauth-up
```

Next, run one or more of the following commands (in separate terminals) to start up the remaining components.
(These commands are the same as the ones inside the `Dockerfile-dev` file for each component.) 

### Flowmachine
```bash
$ cd flowmachine/
$ pipenv run watchmedo auto-restart --recursive --patterns="*.py" --directory="." pipenv run flowmachine
```

### FlowAPI
```bash
$ cd flowapi/
$ pipenv run hypercorn --debug --reload --bind 0.0.0.0:9090 "app.main:create_app()"
```

### FlowAuth

_TODO_


# Verifying the setup

This section provides example commands which you can run to verify that `flowdb`, `flowmachine` and `flowapi` started up
successfully and are wired up correctly so that they can talk to each other.

## Running the integration tests

You can run the integration tests as follows. If these pass, this is a good indication that everything is set up correctly.
```bash
$ cd integretion_tests
$ pipenv install  # only needed once to install dependencies
$ pipenv run pytest
```

## FlowDB

```
$ psql "postgresql://flowdb:flowflow@localhost:9000/flowdb" -c "SELECT flowdb_version()"
```
The output should be similar to this:
```
        flowdb_version
-------------------------------
 (0.2.2-8-g0558488,2019-01-21)
(1 row)
```

## Flowmachine

From within the `flowmachine/` folder, run the following command to send an example message to the Flowmachine server via ZeroMQ.
```
$ pipenv run python -c "from flowmachine.core.server.utils import FM_EXAMPLE_MESSAGE, send_message_and_receive_reply; print(send_message_and_receive_reply(FM_EXAMPLE_MESSAGE))"
```
Expected output:
```
Sending message: {'action': 'run_query', 'query_kind': 'daily_location', 'request_id': 'DUMMY_ID', 'params': {'date': '2016-01-01', 'daily_location_method': 'last', 'aggregation_unit': 'admin3', 'subscriber_subset': 'all'}}
{'status': 'accepted', 'id': 'ddc61a04f608dee16fff0655f91c2057'}
```

## FlowAPI

First navigate to `http://localhost:8080/` (where Flowauth should be running), create a valid token (TODO: add
instructions how to do this or link to the appropriate section in the docs) and store it in the environment
variable `TOKEN`.

Then run the following command to submit a daily location calculation to FlowKit via the API:
```
$ curl -v -X POST http://localhost:9090/api/0/run \
  -H "Authorization: Bearer ${TOKEN}" \
  -H 'Cache-Control: no-cache' \
  -H 'Content-Type: application/json' \
  -d '{"params": {"date":"2016-01-01", "level":"admin3", "daily_location_method":"last", "aggregation_unit":"admin3", "subscriber_subset":"all"}, "query_kind":"daily_location"}'
```

The output should contain a location header that similar to this:
```
[...]
< location: /api/0/poll/ddc61a04f608dee16fff0655f91c2057
[...]
```


CHANGELOG.md
# Changelog
All notable changes to FlowKit will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).

## [Unreleased]

### Added

### Changed

### Fixed

### Removed

## [0.6.0]
### Added
- Added new flowclient API entrypoint, `aggregate_network_objects`, to access equivalent flowmachine query [#601](https://github.com/Flowminder/FlowKit/issues/601)
- FlowAPI now exposes the API spec at the `spec/openapi.json` endpoint, and an interactive version of the spec at the `spec/redoc` endpoint
- Added Makefile target `make up-no_build`, to spin up all containers without building the images
- Added `resync_redis_with_cache` function to cache utils, to allow administrators to align redis with FlowDB [#636](https://github.com/Flowminder/FlowKit/issues/636)
- Added new flowclient API entrypoint, `radius_of_gyration`, to access (with simplified parameters) equivalent flowmachine query `RadiusOfGyration` [#602](https://github.com/Flowminder/FlowKit/issues/602)

### Changed
- The `period` argument to `TotalNetworkObjects` in FlowMachine has been renamed `total_by`
- The `period` argument to `total_network_objects` in FlowClient has been renamed `total_by`
- The `by` argument to `AggregateNetworkObjects` in FlowMachine has been renamed to `aggregate_by`
- The `stop_date` argument to the `modal_location_from_dates` and `meaningful_locations_*` functions in FlowClient has been renamed `end_date` [#470](https://github.com/Flowminder/FlowKit/issues/470) 
- `get_result_by_query_id` now accepts a `poll_interval` argument, which allows polling frequency to be changed
- `RadiusOfGyration` now returns a `value` column instead of an `rog` column
- `TotalNetworkObjects` and `AggregateNetworkObjects` now return a `value` column, rather than `statistic_name`

### Changed
- All environment variables are now in a single `development_environment` file in the project root, development environment setup has been simplified
- Default FlowDB users for FlowMachine and FlowAPI have changed from "analyst" and "reporter" to "flowmachine" and "flowapi", respectively
- Docs and integration tests now use top level compose file
- The following environment variables have been renamed:
  - `FLOWMACHINE_SERVER` (FlowAPI) -> `FLOWMACHINE_HOST`
  - `FM_PASSWORD` (FlowDB), `FLOWDB_PASS` (FlowMachine) -> `FLOWMACHINE_FLOWDB_PASSWORD`
  - `API_PASSWORD` (FlowDB), `FLOWDB_PASS` (FlowAPI) -> `FLOWAPI_FLOWDB_PASSWORD`
  - `FM_USER` (FlowDB), `FLOWDB_USER` (FlowMachine) -> `FLOWMACHINE_FLOWDB_USER`
  - `API_USER` (FlowDB), `FLOWDB_USER` (FlowAPI) -> `FLOWAPI_FLOWDB_USER`
  - `LOG_LEVEL` (FlowMachine) -> `FLOWMACHINE_LOG_LEVEL`
  - `LOG_LEVEL` (FlowAPI) -> `FLOWAPI_LOG_LEVEL`
  - `DEBUG` (FlowDB) -> `FLOWDB_DEBUG`
  - `DEBUG` (FlowMachine) -> `FLOWMACHINE_SERVER_DEBUG_MODE`
- The following Docker secrets have been renamed:
  - `FLOWAPI_DB_USER` -> `FLOWAPI_FLOWDB_USER`
  - `FLOWAPI_DB_PASS` -> `FLOWAPI_FLOWDB_PASSWORD`
  - `FLOWMACHINE_DB_USER` -> `FLOWMACHINE_FLOWDB_USER`
  - `FLOWMACHINE_DB_PASS` -> `FLOWMACHINE_FLOWDB_PASSWORD`
  - `POSTGRES_PASSWORD_FILE` -> `POSTGRES_PASSWORD`
  - `REDIS_PASSWORD_FILE` -> `REDIS_PASSWORD`
- `status` enum in FlowDB renamed to `etl_status`
- `reset_cache` now requires a redis client argument

### Fixed
- Fixed being unable to add new users or servers when running FlowAuth with a Postgres database [#622](https://github.com/Flowminder/FlowKit/issues/622)
- Resetting the cache using `reset_cache` will now reset the state of queries in redis as well [#650](https://github.com/Flowminder/FlowKit/issues/650)
- Fixed `mode` statistic for `AggregateNetworkObjects` [#651](https://github.com/Flowminder/FlowKit/issues/651)

### Removed
- Removed `docker-compose-dev.yml`, and docker-compose files in `docs/`, `flowdb/tests/` and `integration_tests/`.
- Removed `Dockerfile-dev` Dockerfiles
- Removed `ENV` defaults from the FlowMachine Dockerfile
- Removed `POSTGRES_DB` environment variable from FlowDB Dockerfile, database name is now hardcoded as `flowdb`


## [0.5.3]
### Added
- Added new `spatial_aggregate` API endpoint and FlowClient function [#599](https://github.com/Flowminder/FlowKit/issues/599)
- Added new flowclient API entrypoint, total_network_objects(), to access (with simplified parameters) equivalent flowmachine query [#581](https://github.com/Flowminder/FlowKit/issues/581)
- Added new flowclient API entrypoint, location_introversion(), to access (with simplified parameters) equivalent flowmachine query [#577](https://github.com/Flowminder/FlowKit/issues/577)
- Added new flowclient API entrypoint, unique_subscriber_counts(), to access (with simplified parameters) equivalent flowmachine query [#562](https://github.com/Flowminder/FlowKit/issues/562)
- New schema `aggregates` and table `aggregates.aggregates` have been created for maintaining a record of the process and completion of scheduled aggregates.
- New `joined_spatial_aggregate` API endpoint and FlowClient function [#600](https://github.com/Flowminder/FlowKit/issues/600)

### Changed
- `daily_location` and `modal_location` query types are no longer accepted as top-level queries, and must be wrapped using `spatial_aggregate`
- `JoinedSpatialAggregate` no longer accepts positional arguments
- `JoinedSpatialAggregate` now supports "avg", "max", "min", "median", "mode", "stddev" and "variance" stats

### Fixed
- `total_network_objects` no longer returns results from `AggregateNetworkObjects` [#603](https://github.com/Flowminder/FlowKit/issues/603)


## [0.5.2]
### Fixed
- Fixed [#514](https://github.com/Flowminder/FlowKit/issues/514), which would cause the client to hang after submitting a query that couldn't be created
- Fixed [#575](https://github.com/Flowminder/FlowKit/issues/575), so that events at midnight are now considered to be happening on the following day


## [0.5.1]
### Added
- Added `HandsetStats` to FlowMachine.
- Added new `ContactReferenceLocationStats` query class to FlowMachine.

flowapi/flowapi/config.py
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.
import logging
import os
from pathlib import Path


def getsecret(key: str, default: str) -> str:
    """
    Get a value from docker secrets (i.e. read it from a file in
    /run/secrets), return a default if the file is not there.

    Parameters
    ----------
    key: str
        Name of the secret.
    default: str
        Default value to return if the file does not exist

    Returns
    -------
    str
        Value in the file, or default
    """
    try:
        with open(Path("/run/secrets") / key, "r") as fin:
            return fin.read().strip()
    except FileNotFoundError:
        return default


def get_config():
    jwt_secret_key = getsecret("JWT_SECRET_KEY", os.getenv("JWT_SECRET_KEY"))
    log_level = logging.getLevelName(os.getenv("FLOWAPI_LOG_LEVEL", "error").upper())

    flowmachine_host = os.getenv("FLOWMACHINE_HOST", "localhost")
    flowmachine_port = os.getenv("FLOWMACHINE_PORT", "5555")

    flowdb_user = getsecret("FLOWAPI_FLOWDB_USER", os.getenv("FLOWAPI_FLOWDB_USER"))
    flowdb_password = getsecret(
        "FLOWAPI_FLOWDB_PASSWORD", os.getenv("FLOWAPI_FLOWDB_PASSWORD")
    )
    flowdb_host = os.getenv("FLOWDB_HOST")
    flowdb_port = os.getenv("FLOWDB_PORT", 5432)

    return dict(
        JWT_SECRET_KEY=jwt_secret_key,
        FLOWAPI_LOG_LEVEL=log_level,
        FLOWMACHINE_HOST=flowmachine_host,
        FLOWMACHINE_PORT=flowmachine_port,
        FLOWDB_DSN=f"postgres://{flowdb_user}:{flowdb_password}@{flowdb_host}:{flowdb_port}/flowdb",
    )
